proc/msg-nommu.c

```diff
@@ -107,19 +107,23 @@ int proc_recv(u32 port, msg_t *msg, msg_rid_t *rid)
 	}
 
 	kmsg = p->kmessages;
+	if (kmsg != NULL) {
+		kmsg->state = msg_received;
+		LIST_REMOVE(&p->kmessages, kmsg);
+	}
 
 	if (p->closed) {
 		/* Port is being removed */
 		if (kmsg != NULL) {
 			kmsg->state = msg_rejected;
-			LIST_REMOVE(&p->kmessages, kmsg);
 			proc_threadWakeup(&kmsg->threads);
 		}
 
 		hal_spinlockClear(&p->spinlock, &sc);
```

Comment:
I think spinlock clear may be safely moved earlier - `kmsg` is no longer accessible through `p`:
```suggestion
	if (p->closed) {
		/* Port is being removed */
		hal_spinlockClear(&p->spinlock, &sc);
		if (kmsg != NULL) {
			kmsg->state = msg_rejected;
			proc_threadWakeup(&kmsg->threads);
		}
```

Reply:
I kinda agree, but sender still can access `kmsg` if it spuriously wakes up, same reason in lines 129 and 132